flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java

```diff
@@ -1175,6 +1185,31 @@ private void terminateMiniClusterServices() throws Exception {
         }
     }
 
+    /**
+     * Prevent multiple submission of the same JobGraph that has been mutated in between
```

Comment:
Is that really a contract we want to enforce? Why are you concerned with mutating the `JobGraph` between submissions? If I understand it correctly, it would be safe to do so with a proper cluster.

Reply:
Yes it should be as the job graph needs to send over the network so it's effectively cloned, so any mutations to the JobGraph on the client side wouldn't affect the instance used by Dispatcher.

As I've already suggested, the other option would be simply cloning the job graph here.

@zentol do you have any thoughts on this one?

Reply:
The idea here, was to guard test authors for running into the same trap and identify other tests that may suffer from this.

Reply:
I agree that the contract looks strange. Why can not we indeed clone the JobGraph in submitJob?(it is better not to clone but use builder and so but as I understand it is not easy to do so now)

Reply:
Yes, I agree with the analysis. However, I'd rather go with the serialization approach you suggested. Otherwise we are imposing a contract on the MiniCluster that is different from regular clusters. At least that's my take. WDYT?

Reply:
Fixed